### PR TITLE
EA: revert treatment of i8values

### LINKED ISSUE #24615: Accept PandasArray (with correct dtype) in DatetimeArray constructor
I think it make sense that eg in the following case, we extract the numpy array out of a PandasArray, and thus accept it as an argument:

```
In [43]: pd.array([1, 2], dtype='datetime64[ns]')                       
Out[43]: 
<PandasArray>
[numpy.datetime64('1970-01-01T00:00:00.000000001'), numpy.datetime64('1970-01-01T00:00:00.000000002')]
Length: 2, dtype: datetime64[ns]

In [44]: pd.arrays.DatetimeArray(pd.array([1, 2], dtype='datetime64[ns]'))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-44-4cc26981aff2> in <module>
----> 1 pd.arrays.DatetimeArray(pd.array([1, 2], dtype='datetime64[ns]'))

~/scipy/pandas/pandas/core/arrays/datetimes.py in __init__(self, values, dtype, freq, copy)
    265                 "ndarray, or Series or Index containing one of those."
    266             )
--> 267             raise ValueError(msg.format(type(values).__name__))
    268 
    269         if values.dtype == 'i8':

ValueError: Unexpected type 'PandasArray'. 'values' must be a DatetimeArray ndarray, or Series or Index containing one of those.
```

There might be more similar cases like this.

### PR DESCRIPTION
**Pull request by jbrockmendel**
One option for #24559, with some spillover into #24567.

Restores freq validation (needs test)
Retains a _private_ fastpath

- [x] Closes #24615

### CODE REVIEW

**Comment by jbrockmendel** [comment]
No idea how we would go about fixing the failing feather tests.

**Comment by jbrockmendel** [comment]
@TomAugspurger any idea about the feather thing.  AFAICT any change that restores the DTA/Timestamp i8 equivalence will run into this problem.

**Comment by TomAugspurger** [comment]
> AFAICT any change that restores the DTA/Timestamp i8 equivalence will run into this problem.

I think you're correct. On your branch


```python
In [32]: df = pd.DataFrame({"A": pd.date_range('2000', periods=2, tz="US/Central")})

In [33]: df
Out[33]:
                          A
0 2000-01-01 00:00:00-06:00
1 2000-01-02 00:00:00-06:00

In [34]: pa.Table.from_pandas(df).to_pandas()
Out[34]:
                          A
0 2000-01-01 06:00:00-06:00
1 2000-01-02 06:00:00-06:00
```

Which seems strange to me... Why would us changing back to the old way break pyarrow? I'll do some digging.

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -236,83 +236,52 @@ class DatetimeArray(dtl.DatetimeLikeArrayMixin,
...
 
-        if values.dtype == 'i8':
-            # for compat with datetime/timedelta/period shared methods,
-            #  we can sometimes get here with int64 values.  These represent
-            #  nanosecond UTC (or tz-naive) unix timestamps
-            values = values.view(_NS_DTYPE)
-
-        if values.dtype != _NS_DTYPE:
-            msg = (
+        if values.dtype == np.bool_:
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: why do u only check for book here? shouldn’t this. e inside _from_sequeneve?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: well, the good news is that I'm not hallucinating non-existent test cases.  the relevant test was for `to_datetime(False, errors="coerce")`, so I'm moving this check back up to `__init__`.  To be clear, I don't think the three checks in `__init__` should be there at all, but for the moment I'm trying to minimize behavior changes and just get the tests passing.

**Comment by jreback** [comment]
@jbrockmendel this needs additional tests?

**Comment by jbrockmendel** [comment]
> this needs additional tests?

Yes.  Waiting to get weigh-in from @TomAugspurger @jorisvandenbossche @mroeschke.

File: pandas/core/arrays/datetimes.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimes.py]: It's not because `to_datetime` has that behaviour that there should be a check for that here?

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -316,12 +316,12 @@ def _simple_new(cls, values, name=None, freq=None, tz=None, dtype=None):
...
                 dtype = DatetimeTZDtype(tz=tz)
             elif dtype is None:
-                dtype = _NS_DTYPE
+                dtype = values.dtype
 
             values = DatetimeArray(values, freq=freq, dtype=dtype)
             tz = values.tz
             freq = values.freq
-            values = values._data
+            values = values._data.view('i8')
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/datetimes.py]: Why change this to integer data?

(and I see you do a similar change a few times below)

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -2352,7 +2352,7 @@ def _try_coerce_args(self, values, other):
             raise TypeError
         elif is_datetime64_dtype(other):
             # add the tz back
-            other = self._holder(other, dtype=self.dtype)
+            other = self._holder(other.ravel(), dtype=self.dtype)
```
Comment (jorisvandenbossche) [inline pandas/core/internals/blocks.py]: How is this change related to the rest of the PR?

File: pandas/tests/arrays/test_datetimelike.py [hunk 4]
```diff
@@ -340,7 +340,7 @@ def test_from_array_keeps_base(self):
         arr = np.array(['2000-01-01', '2000-01-02'], dtype='M8[ns]')
         dta = DatetimeArray(arr)
 
-        assert dta._data is arr
+        assert dta._data.base is arr
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimelike.py]: why this change?

**Review by jorisvandenbossche — COMMENTED** [review]
I am personally -1 on adding yet another constructor that can parse datetime-like values. We already have `to_datetime` (and DatetimeIndex and to some extent `pd.array` and Series) that can do that. Users that want to parse such values, can use those.

I would personally keep the discussion around the handling of integer values separated from that.

Adding a few comments asking for clarifications on some changes.

File: pandas/tests/arrays/test_datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py]: because a view of `arr` is taken, so the assertion is untrue as-is.

File: pandas/core/indexes/datetimes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: Yep, removing.

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: Here, too, this definitely raised during development without the ravel, but apparently not anymore.  will revert.

File: pandas/tests/arrays/test_datetimelike.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/test_datetimelike.py]: then why was it passing on master? Because you changed it to take a view? Is it important to do that?

Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py]: Because I changed it back to use _from_sequence, which uses sequence_to_dt64ns, which yes takes a view.  And it is important in as much as in the PR implementing sequence_to_dt64ns when the original version only took a view if it was necessary, there was a request to do it unconditionally to save a line

File: pandas/core/arrays/datetimes.py [hunk 5]
```diff
@@ -236,83 +236,51 @@ class DatetimeArray(dtl.DatetimeLikeArrayMixin,
...
         if we are passed a non-dtype compat, then coerce using the constructor
         """
-        dtype = DatetimeTZDtype(tz=tz) if tz else _NS_DTYPE
-
-        return cls(values, freq=freq, dtype=dtype)
+        if tz is not None:
+            # TODO: get tz out of here altogether
+            assert dtype is None
+            tz = timezones.tz_standardize(tz)
+            dtype = DatetimeTZDtype(tz=tz)
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimes.py]: I think this tz normalization needs to be applied to the case of user-provided `dtype=DatetimeTZDtype(...)` as well. I'll try to construct a failing test case.

**Comment by mroeschke** [comment]
I've been out of the loop with DatetimeArray development, but overall I am +1 with the goals outlined in https://github.com/pandas-dev/pandas/issues/24559#issuecomment-451789361

It would be nice if there was 1 entry point to parse all datetime values, but given that we have so many constructors already; this PR's step of maintaining consistency with the others is good.

**Comment by TomAugspurger** [comment]
> I am personally -1 on adding yet another constructor that can parse datetime-like values.

Can you clarify this objection? You would like the following to raise, as it does on master?

```python
In [6]: pd.arrays.DatetimeArray(np.array(['2015-01-01']))
Out[6]:
<DatetimeArray>
['2015-01-01 00:00:00']
Length: 1, dtype: datetime64[ns]
```

**Comment by jreback** [comment]
@jbrockmendel can you rebase

**Comment by jorisvandenbossche** [comment]
> can you clarify this objection?

Yes, I would prefer the situation as it is on master. 
I would strongly prefer to keep the main DatetimeArray constructor simple: only accepting the actual values, and nothing more. (with some exceptions of course, like unboxing containers (Series/Index) that contain such values)

Reasons for my current thinking on that:

- It's not needed. We already have the user-facing `pd.array` entry-point, and `to_datetime` for custom string parsing. 
  If we add similar ability of `pd.array` and `to_datetime` to `DatetimeArray` as well, users will start using it for that, and it is just yet another general constructor that we have to maintain

- Keeping the range of accepted values limited for DatetimeArray gives clearer and more explicit code. If you read this in code, you can know that the values are already an array of datetime64 values. If have something else (scalars, strings, ..), then you can use the specific functionality to deal with those. 
  (although on this aspect: we are still mixing integers and datetimes, and also giving it different behaviour, which already is a bit annoying)

It's basically what we have been discussing regularly before (in https://github.com/pandas-dev/pandas/issues/23212, and probably some other places that I don't directly recall), and what we have done for IntegerArray.

**Comment by TomAugspurger** [comment]
Thanks. There's a 3rd (less important) argument: performance. The slowdown in https://github.com/pandas-dev/pandas/pull/24666 comes down to using `sequence_to_td64ns`. 
https://github.com/pandas-dev/pandas/pull/24666#issuecomment-452331613. That can be resolved, but at the cost of additional code complexity.

**Comment by jbrockmendel** [comment]
This PR is on hold pending discussion in #24559.

There are parts of this PR that I expect to remain unchanged regardless of how that is resolved:
- going through sequence_to_dt64ns rather than re-implementing a bunch of code
  - crippling the public constructor can be done in the block that disallows bool dtype; just disallow a bunch more stuff in that same line.
- doing freq validation and never letting the public constructor return an invalid instance
  - yes I know that has perf implications.  That is why there is a _private_ constructor _simple_new that doesn't do validation

**Comment by TomAugspurger** [comment]
@jbrockmendel the performance in 
https://github.com/pandas-dev/pandas/pull/24666#issuecomment-452331613 is unrelated to freq validation.

**Comment by TomAugspurger** [comment]
Are we holding the RC for this? I haven't followed too closely, but IIRC there's two issues

1. `DTA.__init__` has (incidentally) become more flexible as a result of this PR. That's relatively easy to fix by being bit stricter in the `__init__`
2. The handling of the i8 values?

Where are we on number 2? Do we have agreement on the desired behavior, and does this PR implement that?

**Comment by TomAugspurger** [comment]
On the desired behavior (let's keep the discussion here now that the PR is open):

https://github.com/pandas-dev/pandas/issues/24559#issuecomment-452075476 has a nice summary.

@jorisvandenbossche prefers 

1. Reverting the change to `DTI.__new__` for passing integers. Add a warning that it'll change in the future.
2. Keep `DTA.__init__`  as it is now (same as Timestamp IIRC)

I think that's my preference as well, if I understand things correctly.

**Comment by jorisvandenbossche** [comment]
As far as I understand, this PR is not related to the handling of integers (at least that is what I understood from asking @jbrockmendel about it), so I would prefer to keep the discussion about that in the issue.

**Comment by jorisvandenbossche** [comment]
About this PR, if we want the changes of this PR in general, it's indeed easy to defer the discussion about the flexibility of the constructor by simply adding a check if dtype is M8 or int, and otherwise raising an error.

But I thought there was still disagreement about whether we want this freq validation or not? (because of the performance hit) Or is that resolved?

**Comment by jorisvandenbossche** [comment]
> As far as I understand, this PR is not related to the handling of integers

One thing that it does, though (from https://github.com/pandas-dev/pandas/issues/24559#issuecomment-451789361) is changing the behaviour of DatetimeArray. But so not the reverting of DatetimeIndex + adding a warning there. I would personally keep that for a separate PR, to keep it distinct from the unrelated discussion here (eg flexibility of constructor, freq validation)

**Comment by jbrockmendel** [comment]
I propose to do the following:

- rebase
- revert parts of this PR that affect dtype-treatment, leaving only parts that:
    - ensure freq is validated, so the public DTA constructor never gives an invalid instance
    - retains a private fastpath _simple_new for DTA
- address concerns about the DTI 0.23.4/master behavior

Everyone on board?

**Comment by jorisvandenbossche** [comment]
Not fully on board, that still changes the behaviour of DatetimeArray regarding M8 input, which is something that first should be decided upon (see the table I added in https://github.com/pandas-dev/pandas/issues/24559, writing yet another comment with possible options)

**Comment by jbrockmendel** [comment]
> Not fully on board, that still changes the behaviour of DatetimeArray regarding M8 input,

I'm specifically proposing to revert changes to the behavior of DatetimeArray with regard to M8 input.

**Comment by jbrockmendel** [comment]
> But so without using _from_seqence then? As that does handle M8 differently

The suggested game-plan didn't discuss how to get from here to there.  Just that the DTA behavior would revert to how it in master.
